src/Native/Unix/System.Native/pal_interfaceaddresses.c

```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
     return 0;
 }
 
+int32_t SystemNative_GetNetworkInterfaces(int32_t * interfaceCount, NetworkInterfaceInfo **interfaceList, int32_t * addressCount, IpAddressInfo **addressList )
+{
+    struct ifaddrs* head;   // Pointer to block allocated by getifaddrs().
+    struct ifaddrs* ifaddrsEntry;
+    IpAddressInfo *ai;
+    int count = 0;       // Count of entries returned by getifaddrs().
+    int ip4count = 0;    // Total number of IPv4 addresses.
+    int ip6count = 0;    // Total number of IPv6 addresses.
+    int ifcount = 0;     // Total number of unique network interface.
+    int index;
+    int socketfd = -1;
+
+    NetworkInterfaceInfo *nii;
+
+    if (getifaddrs(&head) == -1)
+    {
+        return -1;
+    }
+
+    ifaddrsEntry = head;
+    while (ifaddrsEntry != NULL)
+    {
+        count ++;
+        if (ifaddrsEntry->ifa_addr != NULL && ifaddrsEntry->ifa_addr->sa_family == AF_INET)
+        {
+            ip4count++;
+        }
+        else if (ifaddrsEntry->ifa_addr != NULL && ifaddrsEntry->ifa_addr->sa_family == AF_INET6)
+        {
+            ip6count++;
+        }
+
+        ifaddrsEntry = ifaddrsEntry->ifa_next;
+    }
+
+    // Allocate estimated space. It can be little bit more than we need.
+    void * memoryBlock = calloc((size_t)count, sizeof(NetworkInterfaceInfo));
+    if (memoryBlock == NULL)
+    {
+        return -1;
+    }
+
+    // Reset head pointers again.
+    ifaddrsEntry = head;
+    *interfaceList = nii = (NetworkInterfaceInfo*)memoryBlock;
+    *addressList = ai = (IpAddressInfo*)(nii + (count - ip4count - ip6count));
+
+    while (ifaddrsEntry != NULL)
+    {
```

Comment:
Can you add comments to each section here highlighting what it's doing?  I'm having trouble following the logic and will be able to review better if I know what each section of code is trying to accomplish.